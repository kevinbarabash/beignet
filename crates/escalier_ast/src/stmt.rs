use generational_arena::Index;

use crate::expr::Expr;
use crate::pattern::Pattern;
use crate::span::Span;
use crate::type_ann::TypeAnn;
use crate::TypeParam;

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum StmtKind {
    Expr {
        expr: Expr,
    },
    Let {
        pattern: Pattern,
        expr: Expr,
        type_ann: Option<TypeAnn>,
    },
    Return {
        arg: Option<Expr>,
    },
    TypeDecl {
        declare: bool,
        name: String,
        type_ann: TypeAnn,
        type_params: Option<Vec<TypeParam>>,
    },
    // TODO:
    // - explicit type annotations
    // - function decls: `fn foo() {}` desugars to `let foo = fn () {}`
    // - class decls: `class Foo {}` desugars to `let Foo = class {}`
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Stmt {
    pub kind: StmtKind,
    pub span: Span,
    pub inferred_type: Option<Index>,
}
