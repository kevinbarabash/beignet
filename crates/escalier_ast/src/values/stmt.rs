use derive_visitor::{Drive, DriveMut};

use crate::values::block::Block;
use crate::values::class::Class;
use crate::values::common::{SourceLocation, Span};
use crate::values::expr::Expr;
use crate::values::ident::Ident;
use crate::values::pattern::Pattern;
use crate::values::type_ann::{TypeAnn, TypeParam};

#[derive(Clone, Debug, Drive, DriveMut, PartialEq, Eq)]
pub struct ClassDecl<T: 'static>
where
    T: Drive + DriveMut,
{
    #[drive(skip)]
    pub ident: Ident, // Why do have `ident` here an in `Class`?
    pub class: Box<Class<T>>,
}

#[derive(Clone, Debug, Drive, DriveMut, PartialEq, Eq)]
pub struct VarDecl<T: 'static>
where
    T: Drive + DriveMut,
{
    pub pattern: Pattern<T>,
    pub type_ann: Option<TypeAnn<T>>,
    pub init: Option<Box<Expr<T>>>,
    #[drive(skip)]
    pub declare: bool,
}

#[derive(Clone, Debug, Drive, DriveMut, PartialEq, Eq)]
pub struct TypeDecl<T: 'static>
where
    T: Drive + DriveMut,
{
    #[drive(skip)]
    pub declare: bool,
    #[drive(skip)]
    pub id: Ident,
    pub type_ann: TypeAnn<T>,
    pub type_params: Option<Vec<TypeParam<T>>>,
}

#[derive(Clone, Debug, Drive, DriveMut, PartialEq, Eq)]
pub struct ForStmt<T: 'static>
where
    T: Drive + DriveMut,
{
    pub pattern: Box<Pattern<T>>,
    pub expr: Box<Expr<T>>,
    pub body: Block<T>,
}

#[derive(Clone, Debug, Drive, DriveMut, PartialEq, Eq)]
pub struct ReturnStmt<T: 'static>
where
    T: Drive + DriveMut,
{
    pub arg: Option<Box<Expr<T>>>,
}

#[derive(Clone, Debug, Drive, DriveMut, PartialEq, Eq)]
pub enum StmtKind<T: 'static>
where
    T: Drive + DriveMut,
{
    // Declarations
    ClassDecl(ClassDecl<T>),
    VarDecl(VarDecl<T>),
    TypeDecl(TypeDecl<T>),

    // Statements
    ExprStmt(Expr<T>),
    ForStmt(ForStmt<T>),
    ReturnStmt(ReturnStmt<T>),
}

#[derive(Clone, Debug, Drive, DriveMut, PartialEq, Eq)]
pub struct Statement<T: 'static>
where
    T: Drive + DriveMut,
{
    #[drive(skip)]
    pub loc: SourceLocation,
    #[drive(skip)]
    pub span: Span,
    pub kind: StmtKind<T>,
    // pub inferred_type: Option<T>,
}
