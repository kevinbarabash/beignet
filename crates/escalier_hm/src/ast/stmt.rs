use generational_arena::Index;

use crate::ast::block::Block;
use crate::ast::class::Class;
use crate::ast::common::{SourceLocation, Span};
use crate::ast::expr::Expr;
use crate::ast::ident::Ident;
use crate::ast::pattern::Pattern;
use crate::ast::type_ann::{TypeAnn, TypeParam};

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ClassDecl {
    pub ident: Ident, // Why do have `ident` here an in `Class`?
    pub class: Box<Class>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct VarDecl {
    pub pattern: Pattern,
    pub type_ann: Option<TypeAnn>,
    pub init: Option<Box<Expr>>,
    pub declare: bool,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct TypeDecl {
    pub declare: bool,
    pub id: Ident,
    pub type_ann: TypeAnn,
    pub type_params: Option<Vec<TypeParam>>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ForStmt {
    pub pattern: Box<Pattern>,
    pub expr: Box<Expr>,
    pub body: Block,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ReturnStmt {
    pub arg: Option<Box<Expr>>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum StmtKind {
    // Declarations
    ClassDecl(ClassDecl),
    VarDecl(VarDecl),
    TypeDecl(TypeDecl),

    // Statements
    ExprStmt(Expr),
    ForStmt(ForStmt),
    ReturnStmt(ReturnStmt),
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Statement {
    pub loc: SourceLocation,
    pub span: Span,
    pub kind: StmtKind,
    pub inferred_type: Option<Index>,
}
