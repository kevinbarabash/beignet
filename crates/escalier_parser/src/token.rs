use crate::expr::Expr;
use crate::jsx::{JSXElement, JSXFragment};
use crate::source_location::SourceLocation;

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum TokenKind {
    Identifier(String), // [a-zA-Z_][a-zA-Z0-9_]*

    // Literals
    BoolLit(bool),
    NumLit(String),
    StrLit(String),
    StrTemplateLit {
        parts: Vec<Token>, // This should only contain StrLit tokens
        exprs: Vec<Expr>,
    },
    Null,
    Undefined,
    JSXElement(JSXElement),
    JSXFragment(JSXFragment),

    // Types
    Number,
    Boolean,
    String,
    Symbol,

    // Keywords
    Let,
    Return,
    Fn,
    Await,
    Async,
    If,
    Else,
    Mut, // denotes a binding to a mutable reference
    Var, // denotes a re-assignable binding
    Match,
    Try,
    Catch,
    Finally,
    Do,

    // Binary Operators
    Plus,
    Minus,
    Times,
    Divide,
    Modulo,
    Equals,
    NotEquals,
    LessThan,
    LessThanOrEqual,
    GreaterThan,
    GreaterThanOrEqual,
    Or,
    And,

    // Assignment Operators
    Assign,
    PlusAssign,
    MinusAssign,
    TimesAssign,
    DivideAssign,
    ModuloAssign,

    // punctuation
    Semicolon,
    Comma,
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    LeftBracket,
    RightBracket,
    Arrow,
    Underscore,
    Question,
    QuestionDot, // used for optional chaining
    Dot,
    DotDot,    // used for ranges
    DotDotDot, // used for rest/spread
    Colon,

    Eof,
}

#[derive(Debug, PartialEq, Eq, Clone)]

pub struct Token {
    pub kind: TokenKind,
    pub loc: SourceLocation,
}
