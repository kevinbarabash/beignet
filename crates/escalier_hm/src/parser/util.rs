use crate::ast::*;

use crate::parser::parse_error::ParseError;
use crate::parser::pattern::parse_pattern;
use crate::parser::type_ann::parse_type_ann;

pub fn text_for_node(node: &tree_sitter::Node, src: &str) -> Result<String, ParseError> {
    match src.get(node.byte_range()) {
        Some(text) => Ok(text.to_owned()),
        None => Err(ParseError::from("error in text_for_node")),
    }
}

pub fn parse_type_params_for_node(
    node: &tree_sitter::Node,
    src: &str,
) -> Result<Option<Vec<TypeParam>>, ParseError> {
    let type_params = match node.child_by_field_name("type_parameters") {
        Some(type_params) => {
            let mut cursor = type_params.walk();
            let type_params = type_params
                .named_children(&mut cursor)
                .into_iter()
                .map(|type_param| {
                    let name_node = type_param.child_by_field_name("name").unwrap();
                    let name = Ident {
                        loc: SourceLocation::from(&name_node),
                        span: name_node.byte_range(),
                        name: text_for_node(&name_node, src)?,
                    };

                    let constraint =
                        if let Some(constraint) = type_param.child_by_field_name("constraint") {
                            Some(Box::from(parse_type_ann(&constraint, src)?))
                        } else {
                            None
                        };

                    let default = if let Some(value) = type_param.child_by_field_name("value") {
                        Some(Box::from(parse_type_ann(&value, src)?))
                    } else {
                        None
                    };

                    Ok(TypeParam {
                        span: type_param.byte_range(),
                        name,
                        constraint,
                        default,
                    })
                })
                .collect::<Result<Vec<_>, ParseError>>()?;
            Some(type_params)
        }
        None => None,
    };

    Ok(type_params)
}

pub fn parse_formal_parameters(
    node: &tree_sitter::Node,
    src: &str,
) -> Result<Vec<EFnParam>, ParseError> {
    assert_eq!(node.kind(), "formal_parameters");

    let mut cursor = node.walk();
    node.named_children(&mut cursor)
        .into_iter()
        .map(|param| {
            let optional = match param.kind() {
                "required_parameter" => false,
                "optional_parameter" => true,
                kind => panic!("Unexpected param kind: {kind}"),
            };
            let pattern = param.child_by_field_name("pattern").unwrap();
            let type_ann = if let Some(type_ann) = param.child_by_field_name("type") {
                Some(parse_type_ann(&type_ann, src)?)
            } else {
                None
            };

            Ok(EFnParam {
                pat: parse_pattern(&pattern, src)?,
                type_ann,
                optional,
            })
        })
        .collect::<Result<Vec<_>, ParseError>>()
}
