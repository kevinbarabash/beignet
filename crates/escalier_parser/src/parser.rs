use std::iter::{Iterator, Peekable};

use crate::source_location::*;
use crate::token::{Token, TokenKind};

pub struct Parser {
    tokens: Peekable<Box<dyn Iterator<Item = Token>>>,
}

impl Parser {
    pub fn new(tokens: Box<dyn Iterator<Item = Token>>) -> Self {
        Self {
            tokens: tokens.peekable(),
        }
    }

    pub fn next(&mut self) -> Token {
        self.tokens.next().unwrap_or(Token {
            kind: TokenKind::Eof,
            loc: SourceLocation {
                start: Position { line: 0, column: 0 },
                end: Position { line: 0, column: 0 },
            },
        })
    }

    pub fn peek(&mut self) -> Token {
        self.tokens.peek().cloned().unwrap_or(Token {
            kind: TokenKind::Eof,
            loc: SourceLocation {
                start: Position { line: 0, column: 0 },
                end: Position { line: 0, column: 0 },
            },
        })
    }
}
