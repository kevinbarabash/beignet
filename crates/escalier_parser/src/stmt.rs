use escalier_ast::values::*;

use crate::expr::parse_expression;
use crate::parse_error::ParseError;
use crate::pattern::*;
use crate::type_ann::parse_type_ann;
use crate::util::*;

pub fn parse_statement(
    node: &tree_sitter::Node,
    src: &str,
) -> Result<Option<Statement>, ParseError> {
    let kind = match node.kind() {
        "lexical_declaration" => {
            return parse_declaration(node, false, src);
        }
        "class_declaration" => {
            return parse_class_decl(node, src);
        }
        "expression_statement" => {
            let expr = node.named_child(0).unwrap();
            let expr = parse_expression(&expr, src)?;
            StmtKind::ExprStmt(expr)
        }
        "ambient_declaration" => {
            let decl = node.named_child(0).unwrap();
            return parse_declaration(&decl, true, src);
        }
        "type_alias_declaration" => {
            let name = node.child_by_field_name("name").unwrap();
            let id = Ident {
                loc: SourceLocation::from(&name),
                span: name.byte_range(),
                name: text_for_node(&name, src)?,
            };
            let type_ann = node.child_by_field_name("value").unwrap();
            let type_ann = parse_type_ann(&type_ann, src)?;

            let type_params = parse_type_params_for_node(node, src)?;

            StmtKind::TypeDecl(TypeDecl {
                declare: false,
                id,
                type_ann,
                type_params,
            })
        }
        "for_in_statement" => {
            let left_node = node.child_by_field_name("left").unwrap();
            let right_node = node.child_by_field_name("right").unwrap();
            let body_node = node.child_by_field_name("body").unwrap();

            StmtKind::ForStmt(ForStmt {
                pattern: Box::from(parse_pattern(&left_node, src)?),
                expr: Box::from(parse_expression(&right_node, src)?),
                body: parse_block_statement(&body_node, src)?,
            })
        }
        "return_statement" => {
            let arg = match node.named_child(0) {
                Some(arg) => Some(Box::from(parse_expression(&arg, src)?)),
                None => None,
            };

            StmtKind::ReturnStmt(ReturnStmt { arg })
        }
        "comment" => {
            return Ok(None); // ignore comments
        }
        "ERROR" => {
            return Err(ParseError::from(format!(
                "failed to parse: '{}'",
                text_for_node(node, src)?
            )))
        }
        _ => return Err(ParseError::from(format!("unhandled: {:#?}", node))),
    };

    Ok(Some(Statement {
        loc: SourceLocation::from(node),
        span: node.byte_range(),
        kind,
    }))

    // $.export_statement,
    // $.import_statement,
    // $.debugger_statement,
    // $.expression_statement,
    // $.declaration,
    // $.statement_block,

    // $.if_statement,
    // $.switch_statement,
    // $.for_statement,
    // $.for_in_statement,
    // $.while_statement,
    // $.do_statement,
    // $.try_statement,
    // $.with_statement,

    // $.break_statement,
    // $.continue_statement,
    // $.return_statement,
    // $.throw_statement,
    // $.empty_statement,
    // $.labeled_statement
}

fn parse_declaration(
    node: &tree_sitter::Node,
    declare: bool,
    src: &str,
) -> Result<Option<Statement>, ParseError> {
    if node.has_error() {
        // TODO: get actual error node so that we can report where the error is
        return Err(ParseError::from("Error parsing declaration"));
    }

    let decl = node.child_by_field_name("decl").unwrap();
    let rec = node.child_by_field_name("rec").is_some();

    let name = decl.child_by_field_name("name").unwrap();
    let pattern = parse_pattern(&name, src)?;

    let type_ann = if let Some(type_ann) = decl.child_by_field_name("type") {
        Some(parse_type_ann(&type_ann, src)?)
    } else {
        None
    };

    let kind = if rec {
        // `let fib = fix((fib) => (n) => ...)`
        // TODO: Fix always wraps a lambda

        let value = decl.child_by_field_name("value").unwrap();

        let lambda_expr = Expr {
            loc: SourceLocation::from(&decl),
            span: decl.byte_range(),
            kind: ExprKind::Lambda(Lambda {
                params: vec![EFnParam {
                    pat: pattern.clone(),
                    type_ann: type_ann.clone(),
                    optional: false,
                }],
                body: BlockOrExpr::Expr(Box::from(parse_expression(&value, src)?)),
                is_async: false,
                return_type: None,
                type_params: None, // TODO: support type params on VarDecls
            }),
            inferred_type: None,
        };

        let fix = Expr {
            loc: SourceLocation::from(&decl),
            span: decl.byte_range(),
            kind: ExprKind::Fix(Fix {
                expr: Box::from(lambda_expr),
            }),
            inferred_type: None,
        };

        StmtKind::VarDecl(VarDecl {
            pattern,
            type_ann,
            init: Some(Box::from(fix)),
            declare,
        })
    } else {
        let init = if let Some(init) = decl.child_by_field_name("value") {
            Some(Box::from(parse_expression(&init, src)?))
        } else {
            None
        };

        StmtKind::VarDecl(VarDecl {
            pattern,
            type_ann,
            init,
            declare,
        })
    };

    Ok(Some(Statement {
        loc: SourceLocation::from(node),
        span: node.byte_range(),
        kind,
    }))
}

fn parse_class_decl(node: &tree_sitter::Node, src: &str) -> Result<Option<Statement>, ParseError> {
    if node.has_error() {
        // TODO: get actual error node so that we can report where the error is
        return Err(ParseError::from("Error parsing declaration"));
    }

    let name_node = node.child_by_field_name("name").unwrap();
    let body_node = node.child_by_field_name("body").unwrap();

    let mut class_members: Vec<ClassMember> = vec![];
    let mut cursor = body_node.walk();
    for child in body_node.named_children(&mut cursor) {
        let kind = child.kind();

        eprintln!("child.kind() = {kind}");

        match kind {
            "method_definition" => {
                let name_node = child.child_by_field_name("name").unwrap();
                let key = Ident {
                    loc: SourceLocation::from(&name_node),
                    span: name_node.byte_range(),
                    name: text_for_node(&name_node, src)?,
                };

                let params = parse_formal_parameters(
                    &child.child_by_field_name("parameters").unwrap(),
                    src,
                )?;
                let body_node = child.child_by_field_name("body").unwrap();
                let body = parse_block_statement(&body_node, src)?;
                let return_type = match child.child_by_field_name("return_type") {
                    Some(type_ann) => Some(parse_type_ann(&type_ann, src)?),
                    None => None,
                };

                let kind = match child.child_by_field_name("kind") {
                    Some(kind) => match text_for_node(&kind, src)?.as_str() {
                        "get" => MethodKind::Getter,
                        "set" => MethodKind::Setter,
                        "*" => todo!(),
                        _ => panic!("Invalid method kind"), // TODO: report an error
                    },
                    None => MethodKind::Method,
                };

                // TODO: add support to the AST for these
                let is_static = child.child_by_field_name("static").is_some();
                let is_mutating = child.child_by_field_name("mut").is_some();

                let type_params = parse_type_params_for_node(&child, src)?;

                if key.name.as_str() == "constructor" {
                    class_members.push(ClassMember::Constructor(Constructor { params, body }));
                } else {
                    class_members.push(ClassMember::Method(ClassMethod {
                        key,
                        kind,
                        lambda: Lambda {
                            params,
                            body: BlockOrExpr::Block(body),
                            is_async: child.child_by_field_name("async").is_some(),
                            return_type,
                            type_params,
                        },
                        is_static,
                        is_mutating,
                    }))
                }
            }
            "public_field_definition" => {
                let name_node = child.child_by_field_name("name").unwrap();
                let key = Ident {
                    loc: SourceLocation::from(&name_node),
                    span: name_node.byte_range(),
                    name: text_for_node(&name_node, src)?,
                };

                let value = match child.child_by_field_name("value") {
                    Some(value) => Some(Box::from(parse_expression(&value, src)?)),
                    None => None,
                };
                let type_ann = match child.child_by_field_name("type") {
                    Some(type_ann) => Some(Box::from(parse_type_ann(&type_ann, src)?)),
                    None => None,
                };

                class_members.push(ClassMember::Prop(ClassProp {
                    key,
                    value,
                    type_ann,
                    is_static: child.child_by_field_name("static").is_some(),
                    is_optional: child.child_by_field_name("optional").is_some(),
                    is_mutable: child.child_by_field_name("mut").is_some(),
                }))
            }
            "private_field_definition" => todo!(),
            _ => todo!(),
        }
    }

    let stmt = Statement {
        loc: SourceLocation::from(node),
        span: node.byte_range(),
        kind: StmtKind::ClassDecl(ClassDecl {
            ident: Ident {
                loc: SourceLocation::from(&name_node),
                span: name_node.byte_range(),
                name: text_for_node(&name_node, src)?,
            },
            class: Box::from(Class {
                ident: Ident {
                    loc: SourceLocation::from(&name_node),
                    span: name_node.byte_range(),
                    name: text_for_node(&name_node, src)?,
                },
                body: class_members,
                type_params: parse_type_params_for_node(node, src)?, // TODO
            }),
        }),
    };

    Ok(Some(stmt))
}

pub fn parse_block_statement(node: &tree_sitter::Node, src: &str) -> Result<Block, ParseError> {
    assert_eq!(node.kind(), "statement_block");

    let mut cursor = node.walk();

    let child_count = node.named_child_count();

    let mut stmts: Vec<Statement> = vec![];

    for (i, child) in node.named_children(&mut cursor).into_iter().enumerate() {
        let is_last = i == child_count - 1;
        if is_last {
            // This is the only place where a named `expression` node
            // should exist.  Everywhere else its contents should be
            // inline.
            if child.kind() == "expression" {
                let expr = child.named_child(0).unwrap();
                let expr = parse_expression(&expr, src)?;
                stmts.push(Statement {
                    loc: SourceLocation::from(&child),
                    span: child.byte_range(),
                    kind: StmtKind::ExprStmt(expr),
                })
            } else if let Some(result) = parse_statement(&child, src)? {
                stmts.push(result);
            }
        } else if let Some(result) = parse_statement(&child, src)? {
            stmts.push(result);
        }
    }

    Ok(Block {
        span: node.byte_range(),
        stmts,
    })
}
