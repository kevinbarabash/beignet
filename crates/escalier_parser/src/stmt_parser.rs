use crate::parser::*;
use crate::span::merge_spans;
use crate::stmt::{Stmt, StmtKind};
use crate::token::*;

impl<'a> Parser<'a> {
    pub fn parse_stmt(&mut self) -> Stmt {
        let token = self.peek().unwrap_or(&EOF).clone();

        match &token.kind {
            TokenKind::Let => {
                self.next().unwrap_or(EOF.clone());
                let pattern = self.parse_pattern();

                let type_ann = match self.peek().unwrap_or(&EOF).kind {
                    TokenKind::Colon => {
                        self.next().unwrap_or(EOF.clone());
                        Some(self.parse_type_ann())
                    }
                    _ => None,
                };

                assert_eq!(self.next().unwrap_or(EOF.clone()).kind, TokenKind::Assign);
                let expr = self.parse_expr();
                assert_eq!(
                    self.next().unwrap_or(EOF.clone()).kind,
                    TokenKind::Semicolon
                );

                let span = merge_spans(&token.span, &expr.get_span());
                Stmt {
                    kind: StmtKind::Let {
                        pattern,
                        expr,
                        type_ann,
                    },
                    span,
                }
            }
            TokenKind::Return => {
                self.next().unwrap_or(EOF.clone());
                let next = self.peek().unwrap_or(&EOF).clone();
                match next.kind {
                    TokenKind::Semicolon => {
                        self.next().unwrap_or(EOF.clone());
                        Stmt {
                            kind: StmtKind::Return { arg: None },
                            span: merge_spans(&token.span, &next.span),
                        }
                    }
                    _ => {
                        let arg = self.parse_expr();
                        assert_eq!(
                            self.next().unwrap_or(EOF.clone()).kind,
                            TokenKind::Semicolon
                        );

                        let span = merge_spans(&next.span, &arg.get_span());
                        Stmt {
                            kind: StmtKind::Return { arg: Some(arg) },
                            span,
                        }
                    }
                }
            }
            _ => {
                eprintln!("--- parse_stmt (expr) ---");
                let expr = self.parse_expr();
                assert_eq!(
                    self.next().unwrap_or(EOF.clone()).kind,
                    TokenKind::Semicolon
                );
                eprintln!("--- parse_stmt (assert semicolon) ---");

                let span = expr.get_span();
                Stmt {
                    kind: StmtKind::Expr { expr },
                    span,
                }
            }
        }
    }

    pub fn parse_program(&mut self) -> Vec<Stmt> {
        let mut stmts = Vec::new();
        while self.peek().unwrap_or(&EOF).kind != TokenKind::Eof {
            stmts.push(self.parse_stmt());
        }
        stmts
    }
}

pub fn parse(input: &str) -> Vec<Stmt> {
    let mut parser = Parser::new(input);
    parser.parse_program()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn single_statement() {
        let input = "let x = 5;";
        let stmts = parse(input);
        assert_eq!(stmts.len(), 1);
    }

    #[test]
    fn single_variable_expression() {
        let input = "x;";
        let stmts = parse(input);
        assert_eq!(stmts.len(), 1);
    }

    #[test]
    fn multiple_statements() {
        let input = r#"
        let x = 5;
        let y = 10;
        x + y;
        return; 
        "#;

        let stmts = parse(input);

        assert_eq!(stmts.len(), 4);
    }

    #[test]
    fn parse_let() {
        insta::assert_debug_snapshot!(parse(r#"let y = m*x + b;"#));
    }

    #[test]
    fn parse_let_with_type_annotation() {
        insta::assert_debug_snapshot!(parse(r#"let y: number = m*x + b;"#));
    }

    #[test]
    fn parse_let_with_destructuring() {
        insta::assert_debug_snapshot!(parse(r#"let {x, y} = point;"#));
    }

    #[test]
    fn parse_let_with_destructuring_and_type_annotation() {
        insta::assert_debug_snapshot!(parse(r#"let {x, y}: Point = point;"#));
    }

    #[test]
    fn parse_assignment() {
        insta::assert_debug_snapshot!(parse(r#"y = m*x + b;"#));
        insta::assert_debug_snapshot!(parse(r#"p.x = 5;"#));
        insta::assert_debug_snapshot!(parse(r#"p["y"] = 10;"#));
    }

    #[test]
    fn parse_conditionals() {
        insta::assert_debug_snapshot!(parse("let max = if (x > y) { x; } else { y; };"));
        insta::assert_debug_snapshot!(parse("if (foo) { console.log(foo); };"));
    }

    #[test]
    fn parse_lambda() {
        insta::assert_debug_snapshot!(parse("let add = fn (x, y) => x + y;"));
        insta::assert_debug_snapshot!(parse("let add = fn (x) => fn (y) => x + y;"));
    }

    #[test]
    fn parse_let_destructuring() {
        insta::assert_debug_snapshot!(parse("let {x, y} = point;"));
        insta::assert_debug_snapshot!(parse("let {x: x1, y: y1} = p1;"));
        insta::assert_debug_snapshot!(parse("let [p1, p2] = line;"));
        insta::assert_debug_snapshot!(parse("let [head, ...tail] = polygon;"));
    }

    #[test]
    fn parse_let_fn_with_fn_type() {
        insta::assert_debug_snapshot!(parse(
            r#"let add: fn (a: number, b: number) => number = fn (a, b) => a + b;"#
        ));
    }
}
