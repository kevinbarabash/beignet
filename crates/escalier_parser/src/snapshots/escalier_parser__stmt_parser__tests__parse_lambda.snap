---
source: crates/escalier_parser/src/stmt_parser.rs
expression: "parse(\"let add = fn (x, y) => x + y\")"
---
[
    Stmt {
        kind: Let {
            is_declare: false,
            is_mut: false,
            pattern: Pattern {
                kind: Ident(
                    BindingIdent {
                        name: "add",
                        span: 4..7,
                        mutable: false,
                    },
                ),
                span: 4..7,
                inferred_type: None,
            },
            expr: Some(
                Expr {
                    kind: Function(
                        Function {
                            type_params: None,
                            params: [
                                FuncParam {
                                    pattern: Pattern {
                                        kind: Ident(
                                            BindingIdent {
                                                name: "x",
                                                span: 14..15,
                                                mutable: false,
                                            },
                                        ),
                                        span: 14..15,
                                        inferred_type: None,
                                    },
                                    type_ann: None,
                                    optional: false,
                                },
                                FuncParam {
                                    pattern: Pattern {
                                        kind: Ident(
                                            BindingIdent {
                                                name: "y",
                                                span: 17..18,
                                                mutable: false,
                                            },
                                        ),
                                        span: 17..18,
                                        inferred_type: None,
                                    },
                                    type_ann: None,
                                    optional: false,
                                },
                            ],
                            body: Expr(
                                Expr {
                                    kind: Binary(
                                        Binary {
                                            left: Expr {
                                                kind: Ident(
                                                    Ident {
                                                        name: "x",
                                                        span: 23..24,
                                                    },
                                                ),
                                                span: 23..24,
                                                inferred_type: None,
                                            },
                                            op: Plus,
                                            right: Expr {
                                                kind: Ident(
                                                    Ident {
                                                        name: "y",
                                                        span: 27..28,
                                                    },
                                                ),
                                                span: 27..28,
                                                inferred_type: None,
                                            },
                                        },
                                    ),
                                    span: 23..28,
                                    inferred_type: None,
                                },
                            ),
                            type_ann: None,
                            is_async: false,
                            is_gen: false,
                        },
                    ),
                    span: 10..28,
                    inferred_type: None,
                },
            ),
            type_ann: None,
        },
        span: 0..28,
        inferred_type: None,
    },
]
